bip-0322.mediawiki

```diff
@@ -0,0 +1,130 @@
+The current message signing standard only works for P2PKH (1...) addresses. By extending it to use a Bitcoin Script based approach, it could be made more generic without causing a too big burden on implementers, who most likely have access to Bitcoin Script interpreters already.
+
+== Specification ==
+
+A new structure <code>SignatureProof</code> is added, which is a simple serializable scriptSig & witnessProgram container.
+
+Two actions "Sign" and "Verify" are defined.
+
+=== SignatureProof container ===
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Type
+!Length
+!Name
+!Comment
+|-
+|Uint32||4||flags||standard flags (1-to-1 with standard flags in Bitcoin Core)
+|-
+|VarInt||1-8||msglen||Number of bytes in message string, excluding NUL termination
+|-
+|Char*||[msglen]||msg||The message being signed for all subjects, excluding NUL termination
+|-
+|Uint8||1||entries||Number of proof entries<ref><strong>Why support multiple proofs?</strong> In particular with proof of funds, it is non-trivial to check a large number of individual proofs (one per UTXO) for duplicates. Software could be written to do so, but it seems more efficient to build this check into the specification itself.</ref>
+|}
+
+The above is followed by [entries] number of signature entries:
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Type
+!Length
+!Name
+!Comment
+|-
+|VarInt||1-8||scriptsiglen||Number of bytes in scriptSig data
+|-
+|Uint8*||[scriptsiglen]||scriptsig||ScriptSig data
+|-
+|VarInt||1-8||witlen||Number of bytes in witness program data
+|-
+|Uint8*||[witlen]||wit||Witness program
+|}
+
+In some cases, the scriptsig may be empty (scriptsiglen=0).
+
+=== Signing ===
+
+The "Sign" action takes as input a scriptPubKey and a message (e.g. "hello world"). It succeeds or fails.
+
+# FAIL if scriptPubKey already exists in scriptPubKeys set, otherwise insert it<ref><strong>Why track duplicates?</strong> Because a 3-entry proof is not proving 3 scriptPubKeys unless they are all distinct, or unless they are proving different UTXO:s (see Future Extensions)</ref>
+# Derive the private key privkey for the scriptPubKey, or FAIL
+# Define the message pre-image as the sequence "Bitcoin Message:" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD)
+# Let sighash = sha256(sha256(scriptPubKey || pre-image))
+# Generate a signature sig with privkey=privkey, sighash=sighash
+
+Repeat the above operation for each scriptPubKey, retaining the scriptPubKeys set. As noted, if the same scriptPubKey appears more than once, the sign operation must fail.
+
+=== Verifying ===
+
+The "Verify" action takes as input a standard flags value, a scriptPubKey, a message, a script sig, and a witness program.
+It emits one of INCONCLUSIVE, VALID, INVALID, or ERROR.
+
+# Return ERROR if scriptPubKey already exists in scriptPubKeys set, otherwise insert it
+# If one or more of the standard flags are unknown, return INCONCLUSIVE
```

Comment:
Isn't it also important for the verifier to return INCONCLUSIVE for an unknown witness version?

Reply:
The standard flags include a "do not allow upgradable flags" flag, which will trigger an error if a future segwit version is used (similar to how it will trigger a warning if a NOP is used).